File: YZScannerObjective-C/YZScanner/Resource/View/YZScannerMaskView.m
//
//  YZScannerPreviewView.m
//  YZScanner
//
//  Created by Broccoli on 2017/1/19.
//  Copyright © 2017年 YZScanner. All rights reserved.
//

#import "YZScannerMaskView.h"

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wobjc-designated-initializers"

@interface YZScannerBackgroundView : UIView

- (instancetype)initWithFrame:(CGRect)frame
           scannerWindowFrame:(CGRect)scannerWindowFrame
              fillColor:(UIColor *)fillColor NS_DESIGNATED_INITIALIZER;

@property (assign, nonatomic) CGRect scannerWindowFrame;
@property (strong, nonatomic) UIColor *fillColor;

@end

@implementation YZScannerBackgroundView

- (instancetype)initWithFrame:(CGRect)frame
           scannerWindowFrame:(CGRect)scannerWindowFrame
              fillColor:(UIColor *)fillColor {
    self = [super initWithFrame:[UIScreen mainScreen].bounds];
    if (!self) {
        return nil;
    }
    self.backgroundColor = [UIColor clearColor];
    self.fillColor = fillColor;
    self.scannerWindowFrame = scannerWindowFrame;
    return self;
}

- (void)drawRect:(CGRect)rect {
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGFloat red,green,blue,alpha;
    BOOL result = [self.fillColor getRed:&red green:&green blue:&blue alpha:&alpha];
    if (!result) {
        return;
    }
    CGContextSetRGBFillColor(context, red, green, blue, alpha);
    //上面
    CGRect fillRect = CGRectMake(0,
                                 0,
                                 self.frame.size.width,
                                 self.scannerWindowFrame.origin.y);
    CGContextFillRect(context, fillRect);
    //左边
    fillRect = CGRectMake(0,
                          self.scannerWindowFrame.origin.y,
                          self.scannerWindowFrame.origin.x,
                          self.scannerWindowFrame.size.height);
    CGContextFillRect(context, fillRect);
    //右边
    fillRect = CGRectMake(self.scannerWindowFrame.origin.x+self.scannerWindowFrame.size.width,
                          self.scannerWindowFrame.origin.y,
                          self.frame.size.width-self.scannerWindowFrame.origin.x-self.scannerWindowFrame.size.width,
                          self.scannerWindowFrame.size.height);
    CGContextFillRect(context, fillRect);
    //下面
    fillRect = CGRectMake(0,
                          self.scannerWindowFrame.origin.y+self.scannerWindowFrame.size.height,
                          self.frame.size.width,
                          self.frame.size.height-(self.scannerWindowFrame.origin.y+self.scannerWindowFrame.size.height));
    CGContextFillRect(context, fillRect);
    
    CGContextStrokePath(context);
}

@end

@interface YZScannerWindowView : UIView

- (instancetype)initWithScannerWindowFrame:(CGRect)scannerWindowFrame
                           borderLineColor:(UIColor *)borderLineColor NS_DESIGNATED_INITIALIZER;

@property (assign, nonatomic) CGRect scannerWindowFrame;
@property (strong, nonatomic) UIColor *borderLineColor;
@property (nonatomic, strong) UIView *referenceLine;

@end

@implementation YZScannerWindowView

- (instancetype)initWithScannerWindowFrame:(CGRect)scannerWindowFrame
                           borderLineColor:(UIColor *)borderLineColor {
    self = [super initWithFrame:scannerWindowFrame];
    if (!self) {
        return nil;
    }
    self.backgroundColor = [UIColor clearColor];
    
    self.borderLineColor = borderLineColor;
    self.scannerWindowFrame = scannerWindowFrame;
    
    self.clipsToBounds = YES;
    [self addSubview:self.referenceLine];
    
    [self addNotification];
    return self;
}

- (void)dealloc {
    [self removeNotification];
}

#pragma mark - Notification
- (void)addNotification {
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(p_applicationWillEnterForeground:) name:UIApplicationWillEnterForegroundNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(p_applicationDidEnterBackground:) name:UIApplicationDidEnterBackgroundNotification object:nil];
}

- (void)removeNotification {
    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIApplicationWillEnterForegroundNotification object:nil];
    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIApplicationDidEnterBackgroundNotification object:nil];
}

- (void)p_applicationWillEnterForeground:(NSNotification*)note {
    _referenceLine.frame = CGRectMake(0, - self.frame.size.height, self.frame.size.width, self.frame.size.height);
    [self startScanning];
}

- (void)p_applicationDidEnterBackground:(NSNotification*)note {
    _referenceLine.frame = CGRectMake(0, - self.frame.size.height, self.frame.size.width, self.frame.size.height);
    [self stopScanning];
}

- (UIView *)referenceLine {
    if (!_referenceLine) {
        _referenceLine = [[UIView alloc] initWithFrame:self.scannerWindowFrame];
        
        CAGradientLayer *gradientLayer = [CAGradientLayer layer];
        gradientLayer.frame = CGRectMake(0, 0, self.scannerWindowFrame.size.width, self.scannerWindowFrame.size.height);
        gradientLayer.colors = @[(id)[UIColor clearColor].CGColor,
                                 (id)[UIColor colorWithRed:0.443 green:0.722 blue:0.569 alpha:1.000].CGColor];
        gradientLayer.locations = @[@(0.5f)];
        gradientLayer.startPoint = CGPointMake(0, 0);
        gradientLayer.endPoint = CGPointMake(0, 1);
        [_referenceLine.layer addSublayer:gradientLayer];
        
        _referenceLine.frame = CGRectMake(0, - _referenceLine.frame.size.height, self.frame.size.width, self.frame.size.height);
    }
    return _referenceLine;
}

- (void)startScanning {
    [UIView animateWithDuration:2.5
                          delay:0
                        options:UIViewAnimationOptionRepeat
                     animations:^{
                         self.referenceLine.hidden = NO;
                         _referenceLine.frame = CGRectMake(0, self.frame.origin.y, self.frame.size.width, self.frame.size.height);
    } completion:nil];
}

- (void)stopScanning {
    [self.referenceLine.layer removeAllAnimations];
    self.referenceLine.hidden = YES;
    _referenceLine.frame = CGRectMake(0, - self.frame.size.height, self.frame.size.width, self.frame.size.height);
}

static NSInteger kLineWidth = 6;
static NSInteger kCornerLength = 18;
static NSInteger kOriginOffset = 0.7;

- (void)drawRect:(CGRect)rect {
    /// 主题色
    CGFloat red,green,blue,alpha;
    BOOL result = [self.borderLineColor getRed:&red green:&green blue:&blue alpha:&alpha];
    if (!result) {
        return;
    }
    
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    CGContextClearRect(context, rect);
    
    CGContextStrokeRect(context, rect);
    CGContextSetRGBStrokeColor(context, red, green, blue, alpha);
    CGContextSetLineWidth(context, 1);
    CGContextAddRect(context, rect);
    CGContextStrokePath(context);
    
    /// 设置 线宽
    CGContextSetLineWidth(context, kLineWidth);
   
    CGContextSetRGBStrokeColor(context, red, green, blue, alpha);
    
    /// 左上角
    CGPoint pointsTopLeftX[] = {
        CGPointMake(CGRectGetMinX(rect) + kCornerLength,
                    CGRectGetMinY(rect) + kOriginOffset),
        CGPointMake(CGRectGetMinX(rect),
                    CGRectGetMinY(rect) + kOriginOffset)
    };
    CGPoint pointsTopLeftY[] = {
        CGPointMake(rect.origin.x + kOriginOffset,
                    rect.origin.y + kCornerLength),
        CGPointMake(rect.origin.x + kOriginOffset,
                    rect.origin.y)
    };
    
    CGContextAddLines(context, pointsTopLeftX, 2);
    CGContextAddLines(context, pointsTopLeftY, 2);
    

    /// 左下角
    CGPoint pointsLeftBottomX[] = {
        CGPointMake(CGRectGetMinX(rect),
                    CGRectGetMaxY(rect) - kOriginOffset),
        CGPointMake(CGRectGetMinX(rect) + kCornerLength + kOriginOffset,
                    CGRectGetMaxY(rect) - kOriginOffset)
    };
    CGPoint pointsLeftBottomY[] = {
        CGPointMake(CGRectGetMinX(rect) + kOriginOffset,
                                               CGRectGetMaxY(rect)),
        CGPointMake(CGRectGetMinX(rect) + kOriginOffset,
                    CGRectGetMaxY(rect) - kCornerLength)
    };
    
    CGContextAddLines(context, pointsLeftBottomX, 2);
    CGContextAddLines(context, pointsLeftBottomY, 2);

    /// 右上角
    CGPoint pointsRightTopX[] = {
        CGPointMake(CGRectGetMaxX(rect) - kCornerLength,
                    CGRectGetMinY(rect) + kOriginOffset),
        CGPointMake(CGRectGetMaxX(rect),
                    CGRectGetMinY(rect) + kOriginOffset)
    };
    CGPoint pointsRightTopY[] = {
        CGPointMake(CGRectGetMaxX(rect) - kOriginOffset,
                                             CGRectGetMinY(rect)),
        CGPointMake(CGRectGetMaxX(rect) - kOriginOffset,
                    CGRectGetMinY(rect) + kCornerLength + kOriginOffset)
    };
    
    CGContextAddLines(context, pointsRightTopX, 2);
    CGContextAddLines(context, pointsRightTopY, 2);

    // 右下角
    CGPoint pointsRightBottomX[] = {
        CGPointMake(CGRectGetMaxX(rect) - kCornerLength,
                    CGRectGetMaxY(rect) - kOriginOffset),
        CGPointMake(CGRectGetMaxX(rect),
                    CGRectGetMaxY(rect) - kOriginOffset)
    };
    CGPoint pointsRightBottomY[] = {
        CGPointMake(CGRectGetMaxX(rect) - kOriginOffset,
                    CGRectGetMaxY(rect) - kCornerLength),
        CGPointMake(CGRectGetMaxX(rect) - kOriginOffset,
                    CGRectGetMaxY(rect))
    };
    
    CGContextAddLines(context, pointsRightBottomX, 2);
    CGContextAddLines(context, pointsRightBottomY, 2);
    
    CGContextStrokePath(context);
}

@end

@interface YZScannerMaskView ()

@property (nonatomic, assign) CGRect scannerWindowFrame;
@property (nonatomic, strong) YZScannerBackgroundView *backgroundView;
@property (nonatomic, strong) YZScannerWindowView *windowView;
@end

@implementation YZScannerMaskView

#pragma mark - Initializing
- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (!self) {
        return nil;
    }
    self.backgroundColor = [UIColor clearColor];
    [self setupSubviews];
    return self;
}

#pragma mark - setup subviews
- (void)setupSubviews {
    [self addSubview:self.backgroundView];
    [self addSubview:self.windowView];
    [self startScanning];
}

#pragma mark - getter
- (YZScannerBackgroundView *)backgroundView {
    if (!_backgroundView) {
        _backgroundView = [[YZScannerBackgroundView alloc] initWithFrame:self.frame
                                                      scannerWindowFrame:self.scannerWindowFrame
                                                               fillColor:[[UIColor blackColor] colorWithAlphaComponent:0.6]];
    }
    return _backgroundView;
}

- (YZScannerWindowView *)windowView {
    if (!_windowView) {
        _windowView = [[YZScannerWindowView alloc] initWithScannerWindowFrame:self.scannerWindowFrame
                                                              borderLineColor:[UIColor colorWithRed:0.443 green:0.722 blue:0.569 alpha:1.0]];
    }
    return _windowView;
}

- (CGRect)scannerWindowFrame {
    if (CGRectEqualToRect(_scannerWindowFrame, CGRectZero)) {
        CGFloat y = 175;
        CGFloat x = 25;
        CGFloat width = [UIScreen mainScreen].bounds.size.width - 2 * x;
        CGFloat height = 220;
        _scannerWindowFrame = CGRectMake(x, y, width, height);
    }
    return _scannerWindowFrame;
}

#pragma mark - publick method
- (void)startScanning {
    [self.windowView startScanning];
}

- (void)stopScanning {
    [self.windowView stopScanning];
}

const NSInteger kActivityBackgroundViewTag = 1001;
const NSInteger kActivityViewTag = 1002;

- (void)startLoadingAnimation {
    self.windowView.hidden = YES;
    self.backgroundView.hidden = YES;
    
    UIView *activityBackgroundView = [[UIView alloc] initWithFrame:self.frame];
    activityBackgroundView.backgroundColor = [[UIColor blackColor] colorWithAlphaComponent: 0.6];
    activityBackgroundView.tag = kActivityBackgroundViewTag;
    [self addSubview:activityBackgroundView];
    
    UIActivityIndicatorView *activityView = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhiteLarge];
    activityView.center = self.center;
    activityView.tag = kActivityViewTag;
    [self addSubview:activityView];
    [activityView startAnimating];
}

- (void)stopLoadingAnimation {
    self.windowView.hidden = NO;
    self.backgroundView.hidden = NO;
    
    UIView *activityBackgroundView = [self viewWithTag:kActivityBackgroundViewTag];
    [activityBackgroundView removeFromSuperview];
    UIView *activityView = [self viewWithTag:kActivityViewTag];
    [activityView removeFromSuperview];
}

@end

#pragma clang diagnostic pop

File: YZScannerSwift/QRCode/Resource/YZScannerMaskView.swift
//
//  YZScannerMaskView.swift
//  QRCode
//
//  Created by Broccoli on 2017/2/16.
//  Copyright © 2017年 Broccoli. All rights reserved.
//

import UIKit

class YZScannerMaskView: UIView {

    var scannerWindowFrame = CGRect.zero
    var windowView: YZScannerWindowView!
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.backgroundColor = UIColor.clear
        self.setupSubviews()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    // MARK: - setup subviews
    func setupSubviews() {
        self.addSubview(self.backgroundView)
        self.addSubview(self.windowView)
        self.startScanning()
    }
    
    // MARK: - getter
    let backgroundView: YZScannerBackgroundView = {
        if !self.backgroundView {
            self.backgroundView = YZScannerBackgroundView(frame: self.frame, scannerWindowFrame: self.scannerWindowFrame, fillColor: UIColor.black.withAlphaComponent(0.6))
        }
        return self.backgroundView
    }()
    
    func windowView() -> YZScannerWindowView {
        if !self.windowView {
            self.windowView = YZScannerWindowView(self.scannerWindowFrame, borderLineColor: UIColor(red: CGFloat(0.443), green: CGFloat(0.722), blue: CGFloat(0.569), alpha: CGFloat(1.0)))
        }
        return self.windowView
    }
    
    func scannerWindowFrame() -> CGRect {
        if self.scannerWindowFrame.equalTo(CGRect.zero) {
            var y: CGFloat = 175
            var x: CGFloat = 25
            var width: CGFloat = UIScreen.main.bounds.size.width - 2 * x
            var height: CGFloat = 220
            self.scannerWindowFrame = CGRect(x: x, y: y, width: width, height: height)
        }
        return self.scannerWindowFrame
    }
    // MARK: - publick method
    
    func startScanning() {
        self.windowView().startScanning()
    }
    
    func stopScanning() {
        self.windowView().stopScanning()
    }
    let kActivityBackgroundViewTag: Int = 1001
    let kActivityViewTag: Int = 1002
    
    func startLoadingAnimation() {
        self.windowView.isHidden = true
        self.backgroundView.isHidden = true
        var activityBackgroundView = UIView(frame: self.frame)
        activityBackgroundView.backgroundColor = UIColor.black.withAlphaComponent(0.6)
        activityBackgroundView.tag = kActivityBackgroundViewTag
        self.addSubview(activityBackgroundView)
        var activityView = UIActivityIndicatorView(activityIndicatorStyle: .whiteLarge)
        activityView.center = self.center
        activityView.tag = kActivityViewTag
        self.addSubview(activityView)
        activityView.startAnimating()
    }
    
    func stopLoadingAnimation() {
        self.windowView.isHidden = false
        self.backgroundView.isHidden = false
        var activityBackgroundView: UIView? = self.viewWithTag(kActivityBackgroundViewTag)
        activityBackgroundView?.removeFromSuperview()
        var activityView: UIView? = self.viewWithTag(kActivityViewTag)
        activityView?.removeFromSuperview()
    }
}

class YZScannerBackgroundView: UIView {
    
    var scannerWindowFrame: CGFloat
    var fillColor: UIColor
    
    override init(frame: CGRect, scannerWindowFrame: CGRect, fill fillColor: UIColor) {
        super.init(frame: UIScreen.main.bounds)
        self.backgroundColor = UIColor.clear
        self.fillColor = fillColor
        self.scannerWindowFrame = scannerWindowFrame
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func draw(_ rect: CGRect) {
        var context: CGContext? = UIGraphicsGetCurrentContext()
        var red: CGFloat
        var green: CGFloat
        var blue: CGFloat
        var alpha: CGFloat
        var result: Bool = self.fillColor.getRed(red, green: green, blue: blue, alpha: alpha)
        if !result {
            return
        }
        context.setFillColor(red: red, green: green, blue: blue, alpha: alpha)
        //上面
        var fillRect = CGRect(x: CGFloat(0), y: CGFloat(0), width: CGFloat(self.frame.size.width), height: CGFloat(self.scannerWindowFrame.origin.y))
        context.fill(fillRect)
        //左边
        fillRect = CGRect(x: CGFloat(0), y: CGFloat(self.scannerWindowFrame.origin.y), width: CGFloat(self.scannerWindowFrame.origin.x), height: CGFloat(self.scannerWindowFrame.size.height))
        context.fill(fillRect)
        //右边
        fillRect = CGRect(x: CGFloat(self.scannerWindowFrame.origin.x + self.scannerWindowFrame.size.width), y: CGFloat(self.scannerWindowFrame.origin.y), width: CGFloat(self.frame.size.width - self.scannerWindowFrame.origin.x - self.scannerWindowFrame.size.width), height: CGFloat(self.scannerWindowFrame.size.height))
        context.fill(fillRect)
        //下面
        fillRect = CGRect(x: CGFloat(0), y: CGFloat(self.scannerWindowFrame.origin.y + self.scannerWindowFrame.size.height), width: CGFloat(self.frame.size.width), height: CGFloat(self.frame.size.height - (self.scannerWindowFrame.origin.y + self.scannerWindowFrame.size.height)))
        context.fill(fillRect)
        context.strokePath()
    }
}

class YZScannerWindowView {
    
    var scannerWindowFrame = CGRect.zero
    var borderLineColor: UIColor!
    var referenceLine: UIView!
    
    override init(scannerWindowFrame: CGRect, borderLineColor: UIColor) {
        super.init(frame: scannerWindowFrame)
        self.backgroundColor = UIColor.clear
        self.borderLineColor = borderLineColor
        self.scannerWindowFrame = scannerWindowFrame
        self.clipsToBounds = true
        self.addSubview(self.referenceLine)
        self.addNotification()
    }
    
    deinit {
        self.removeNotification()
    }
    // MARK: - Notification
    
    func addNotification() {
        NotificationCenter.default.addObserver(self, selector: #selector(self.p_applicationWillEnterForeground), name: NSNotification.Name.UIApplicationWillEnterForeground, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(self.p_applicationDidEnterBackground), name: NSNotification.Name.UIApplicationDidEnterBackground, object: nil)
    }
    
    func removeNotification() {
        NotificationCenter.default.removeObserver(self, name: NSNotification.Name.UIApplicationWillEnterForeground, object: nil)
        NotificationCenter.default.removeObserver(self, name: NSNotification.Name.UIApplicationDidEnterBackground, object: nil)
    }
    
    @objc func p_applicationWillEnterForeground(_ note: Notification) {
        self.referenceLine.frame = CGRect(x: CGFloat(0), y: CGFloat(-self.frame.size.height), width: CGFloat(self.frame.size.width), height: CGFloat(self.frame.size.height))
        self.startScanning()
    }
    
    @objc func p_applicationDidEnterBackground(_ note: Notification) {
        self.referenceLine.frame = CGRect(x: CGFloat(0), y: CGFloat(-self.frame.size.height), width: CGFloat(self.frame.size.width), height: CGFloat(self.frame.size.height))
        self.stopScanning()
    }
    
    //  Converted with Swiftify v1.0.6242 - https://objectivec2swift.com/
    
    func referenceLine() -> UIView? {
        if !self.referenceLine {
            self.referenceLine = UIView(frame: self.scannerWindowFrame)
            var gradientLayer = CAGradientLayer()
            gradientLayer.frame = CGRect(x: CGFloat(0), y: CGFloat(0), width: CGFloat(self.scannerWindowFrame.size.width), height: CGFloat(self.scannerWindowFrame.size.height))
            gradientLayer.colors = [(UIColor.clear.cgColor as? Any), (UIColor(red: CGFloat(0.443), green: CGFloat(0.722), blue: CGFloat(0.569), alpha: CGFloat(1.000)).cgColor as? Any)]
            gradientLayer.locations = [(0.5)]
            gradientLayer.startPoint = CGPoint(x: CGFloat(0), y: CGFloat(0))
            gradientLayer.endPoint = CGPoint(x: CGFloat(0), y: CGFloat(1))
            self.referenceLine.layer.addSublayer(gradientLayer)
            self.referenceLine.frame = CGRect(x: CGFloat(0), y: CGFloat(-self.referenceLine.frame.size.height), width: CGFloat(self.frame.size.width), height: CGFloat(self.frame.size.height))
        }
        return self.referenceLine
    }
    
    func startScanning() {
        UIView.animate(withDuration: 2.5, delay: 0, options: .repeat, animations: {() -> Void in
            self.referenceLine.isHidden = false
            self.referenceLine.frame = CGRect(x: CGFloat(0), y: CGFloat(self.frame.origin.y), width: CGFloat(self.frame.size.width), height: CGFloat(self.frame.size.height))
        }, completion: { _ in })
    }
    
    func stopScanning() {
        self.referenceLine().layer.removeAllAnimations()
        self.referenceLine.isHidden = true
        self.referenceLine.frame = CGRect(x: CGFloat(0), y: CGFloat(-self.frame.size.height), width: CGFloat(self.frame.size.width), height: CGFloat(self.frame.size.height))
    }
    var kLineWidth: Int = 6
    var kCornerLength: Int = 18
    var kOriginOffset: CGFloat = 0.7
    
    //  The converted code is limited by 2 KB.
    //  Upgrade your plan to remove this limitation.
    
    //  Converted with Swiftify v1.0.6242 - https://objectivec2swift.com/
    
    override func draw(_ rect: CGRect) {
        /// 主题色
        var red: CGFloat
        var green: CGFloat
        var blue: CGFloat
        var alpha: CGFloat
        var result: Bool = self.borderLineColor.getRed(red, green: green, blue: blue, alpha: alpha)
        if !result {
            return
        }
        guard let context = UIGraphicsGetCurrentContext() else {
            return
        }
        context.clear(rect)
        context.stroke(rect)
        context.setStrokeColor(red: red, green: green, blue: blue, alpha: alpha)
        context.setLineWidth(CGFloat(1))
        context.addRect(rect)
        context.strokePath()
        /// 设置 线宽
        context.setLineWidth(CGFloat(kLineWidth))
        context.setStrokeColor(red: red, green: green, blue: blue, alpha: alpha)
        /// 左上角
        var pointsTopLeftX: [CGPoint] = [CGPoint(x: CGFloat(rect.minX + kCornerLength), y: CGFloat(rect.minY + kOriginOffset)), CGPoint(x: CGFloat(rect.minX), y: CGFloat(rect.minY + kOriginOffset))]
        var pointsTopLeftY: [CGPoint] = [CGPoint(x: CGFloat(rect.origin.x + kOriginOffset), y: CGFloat(rect.origin.y + kCornerLength)), CGPoint(x: CGFloat(rect.origin.x + kOriginOffset), y: CGFloat(rect.origin.y))]
        context.addLines(between: pointsTopLeftX)
        context.addLines(between: pointsTopLeftY)
        /// 左下角
        var pointsLeftBottomX: [CGPoint] = [CGPoint(x: CGFloat(rect.minX), y: CGFloat(rect.maxY - kOriginOffset)), CGPoint(x: CGFloat(rect.minX + kCornerLength + kOriginOffset), y: CGFloat(rect.maxY - kOriginOffset))]
        var pointsLeftBottomY: [CGPoint] = [CGPoint(x: CGFloat(rect.minX + kOriginOffset), y: CGFloat(rect.maxY)), CGPoint(x: CGFloat(rect.minX + kOriginOffset), y: CGFloat(rect.maxY - kCornerLength))]
        context.addLines(between: pointsLeftBottomX)
        context.addLines(between: pointsLeftBottomY)
        /// 右上角
        var pointsRightTopX: [CGPoint] = [CGPoint(x: CGFloat(rect.maxX - kCornerLength), y: CGFloat(rect.minY + kOriginOffset)), CGPoint(x: CGFloat(rect.maxX), y: CGFloat(rect.minY + kOriginOffset))]
        var pointsRightTopY: [CGPoint] = [CGPoint(x: CGFloat(rect.maxX - kOriginOffset), y: CGFloat(rect.minY)), CGPoint(x: CGFloat(rect.maxX - kOriginOffset), y: CGFloat(rect.minY + kCornerLength + kOriginOffset))]
        context.addLines(between: pointsRightTopX)
        context.addLines(between: pointsRightTopY)
        // 右下角
        var pointsRightBottomX: [CGPoint] = [CGPoint(x: CGFloat(rect.maxX - kCornerLength), y: CGFloat(rect.maxY - kOriginOffset)), CGPoint(x: CGFloat(rect.maxX), y: CGFloat(rect.maxY - kOriginOffset))]
        var pointsRightBottomY: [CGPoint] = [CGPoint(x: CGFloat(rect.maxX - kOriginOffset), y: CGFloat(rect.maxY - kCornerLength)), CGPoint(x: CGFloat(rect.maxX - kOriginOffset), y: CGFloat(rect.maxY))]
        context.addLines(between: pointsRightBottomX)
        context.addLines(between: pointsRightBottomY)
        context.strokePath()
    }
}
